ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,179 @@
+	r := &taskManifestResponder{
+		taskComparer:      taskComparer,
+		snAccessor:        snAccessor,
+		messageIDAccessor: messageIDAccessor,
+		cluster:           cluster,
+		metricsFactory:    metricsFactory,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*taskManifestResponder) Name() string { return "task manifest responder" }
+
+func (tmr *taskManifestResponder) HandlerFunc() wsclient.RequestHandler {
+	return tmr.handleTaskManifestMessage
+}
+
+// handleTaskManifestMessage is the high level caller to handle a task manifest message from ACS.
+// It will kick off the call stack of processTaskManifestMessage calling ack and send TaskManifestMessage
+func (tmr *taskManifestResponder) handleTaskManifestMessage(message *ecsacs.TaskManifestMessage) error {
+	logger.Info(fmt.Sprintf("Processing %s with MessageId %s", TaskManifestMessageName, aws.StringValue(message.MessageId)))
+	m := tmr.metricsFactory.New(metrics.TaskManifestHandlingDuration)
+	err := tmr.processTaskManifestMessage(message)
+	m.Done(err)
+	return err
+}
+
+// processTaskManifestMessage processes a task manifest message from ACS. It verifies
+// whether a task manifest message is stale or not. If not, it proceeds to create a
+// list of stop candidates.
+func (tmr *taskManifestResponder) processTaskManifestMessage(
+	message *ecsacs.TaskManifestMessage) error {
+	manifestSeqNum := aws.Int64Value(message.Timeline)
+	agentLatestSeqNum := tmr.snAccessor.GetLatestSequenceNumber()
+
+	// Verify that task manifest isn't stale.
+	// The agent will keep track of the highest sequence number received in both PayloadMessage and TaskManifestMessage.
+	// If the sequence number on TaskManifestMessage is lower to the latest known one, the TaskManifestMessage must be discarded.
+	// ACS will guarantee that the sequence number is indeed increasing and valid for this use case.
+	if manifestSeqNum < agentLatestSeqNum {
+		logger.Warn(fmt.Sprintf("Discarding task manifest message. "+
+			"Sequence number from manifest is less than Agent latest sequence number."), logger.Fields{
+			"ManifestSeqNum":    manifestSeqNum,
+			"AgentLatestSeqNum": agentLatestSeqNum,
+		})
+		return nil
+	}
+
+	// Save the task manifest's messageID and sequence number so that it can be compared
+	// by the taskStopVerificationAckResponder.
+	err := tmr.messageIDAccessor.SetMessageID(aws.StringValue(message.MessageId))
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest messageID")
+	}
+	err = tmr.snAccessor.SetLatestSequenceNumber(message)
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest sequence number")
+	}
+
+	// Get list of tasks on instance.
+	tasksToMessage, err := tmr.taskComparer.CompareRunningTasksOnInstanceWithManifest(message)
+	if err != nil {
+		return err
+	}
+
+	// Acknowledge the task manifest message asynchronously.
+	var wg sync.WaitGroup
```

Comment:
Why do we need to wait for these function calls finish before returning? It seems like using WaitGroup here is unnecessary.

Reply:
Digging into the code history on this one, it was implemented at the time of writing these responders as we want to force the `TaskStopVerificationAckResponder` to wait until we have sent both of these messages. See diagram [here](https://quip-amazon.com/DG3FAbfXYZid/Full-Task-Sync-for-Fargate-Agent#CfN9CAhSuuZ). Per Amogh's comment earlier I am unsure if it matters as we may be using a single TCP connection. Also unsure if acs will handle messages asynchronously, or wait for one message to be sent before another can be received.

I do not think it is asynchronous, so I do not think we need these goroutines here.

Reply:
Task stop verification ACK responder does not start running/processing anything until a task stop verification ACK has been sent back to us from ACS anyway (i.e., until we have sent a task stop verification message to ACS, they have received it and processed it successfully). From what I understand, waiting here in the task manifest responder does not force the `TaskStopVerificationAckResponder` to do anything and it is already enforced that task stop verification ACK responder does not process a task stop verification ACK before the corresponding task stop verification message is sent by Agent to ACS.

 > Also unsure if acs will handle messages asynchronously
Session/ACS handler consumes messages in the order that they are received (i.e., new goroutines are not spawned on a per message basis), but even if ACS handled messages asynchronously I believe the previous/above point still stands.

Goroutines themselves could still be beneficial here such that we do not block on these calls and allow to proceed with processing other ACS messages before they complete (i.e., current behavior). But I imagine the benefit is marginal here since these calls are short lived anyway. Similar comment on this/question about this here: https://github.com/aws/amazon-ecs-agent/pull/3844/files#r1293959305

Reply:
Synced offline with Amogh and we will include the waitgroup after all.

Reply:
Synced offline with Tim - the purpose of using wait groups here is to make sure that both calls to `ackTaskManifestMessage` and `sendTaskStopVerification` have completed before the call to `m.Done(err)` in `handleTaskManifestMessage` function.

Reply:
Oh I didn't focus much on the wait part. I was just curious why we call `ackTaskManifestMessage` and `sendTaskStopVerification` asynchronously.